### ISSUE TITLE
Migrate the Maven layout provider tests to use the new annotations

### ISSUE THREAD

**Issue by carlspring** [strongbox/strongbox#1165]
# Task Description

We recently implemented #1083 and #1084 and we now need to re-work tests listed below to use these annotations.

Please, note:
* This task is quite large, but the intention is for it to be done in chunks -- you can fix one, or more tests and create a pull request and link it to this task.
* Multiple people can work on this task. For this to work out properly, please add a comment, when you decide to pick some test(s) to work on, so that others know not to work on them.
* This task is good for new joiners who have sufficient experience with JUnit and testing and will quickly help them learn more about the project.

# What Needs To Be Done

* There should be no creation of repositories and test resources (such as artifacts) in initialization methods annotated with `@BeforeAll`, or `@BeforeEach`, such as, for example, `initialize()` methods.
* At the moment, the test methods are all creating their own test repositories and artifacts. The code for this will have to be dropped and replaced with the annotations.
* Drop the `removeRepositories()` methods.
* Drop the `cleanUp()` methods.
* ~Drop all fields that are names of storages and repositories. No two test methods/classes should share the same names of repositories, or artifacts.~ (@carlspring annotations have synchronization based on artifact and repository ID, so it should be safe not to use unique names this will only affect on concurrency factor)
* Each test method needs to generate its own test repository (or repositories) and test resources (typically, artifacts).

For an illustration of a properly migrated test, please have a look at the `RepositoryManagementTest`.

## To create repositories using annotations:
```
@ExtendWith(RepositoryManagementTestExecutionListener.class)
@Test
public void testWithRepository(@TestRepository(storageId = "storage0",
                                               repositoryId = "repo1",
                                               layout = MavenArtifactCoordinates.LAYOUT_NAME) 
                               Repository repository)
{
    System.out.println(repository.getId());
}
```
###### Note
* _`storage = "storage0"` is default value, so it can be omitted_

## To create indexed Maven repository using annotations:
```
@ExtendWith(RepositoryManagementTestExecutionListener.class)
@Test
public void testWithRepository(@TestRepository(repository = "indexed_repo1",
                                               layout = MavenArtifactCoordinates.LAYOUT_NAME,
                                               setup = MavenIndexedRepositorySetup.class) 
                               Repository repository)
{
    System.out.println(repository.getId());
}
```
## To create test artifacts using annotations:
```
@ExtendWith({ RepositoryManagementTestExecutionListener.class,
              ArtifactManagementTestExecutionListener.class })
@Test
public void testArtifact(@TestRepository(layout = MavenArtifactCoordinates.LAYOUT_NAME,
                                          repository = "r1")
                         Repository r1,
                         @TestArtifact(resource = "org/carlspring/strongbox/test/test-artifact/1.0/test-artifact-1.0.jar",
                                       generator = MavenArtifactGenerator.class)
                         Path standaloneArtifact,
                         @TestArtifact(repository = "r1",
                                       resource = "org/carlspring/strongbox/test/test-artifact/1.0/test-artifact-1.0.jar",
                                       generator = MavenArtifactGenerator.class)
                         Path repositoryArtifact)
{
...
}
```
###### Note
* _`@TestArtifact` will deploy only with `repository` attribute, without this attribute it will just generate artifacts without deploy_

## To create multiple test artifact versions 
```
@ExtendWith({ RepositoryManagementTestExecutionListener.class,
              ArtifactManagementTestExecutionListener.class })
@Test
public void testArtifact(@TestRepository(layout = MavenArtifactCoordinates.LAYOUT_NAME,
                                          repository = "r1")
                         Repository r1,
                         @TestArtifact(repository = "r1",
                                       id = "org.carlspring.strongbox.test:test-artifact", 
                                       versions = {"1.0", "2.0"}
                                       generator = MavenArtifactGenerator.class)
                         List<Path> artifacts)
{
...
}
```
###### Note
* _In case of `id` and `versions` you should use `List<Path>` as test method parameter, so the `List` will contain ordered artifact versions. In above example you can get artifact with version `1.0` using `artifacts.get(0)`_

# Task List

The following tests will need to be migrated:

* [x]  `org.carlspring.strongbox.services.ConfigurationManagementServiceImplTest` (@ptirador)
* [x] `org.carlspring.strongbox.services.ArtifactSearchServiceImplTest`
* [x] `org.carlspring.strongbox.services.ArtifactMetadataServiceSnapshotsTest`(@ptirador)
* [x] `org.carlspring.strongbox.services.ArtifactManagementServiceImplTest` @yugander-krishan-singh 
* [x] `org.carlspring.strongbox.services.MavenRepositoryManagementServiceImplTest` @yugander-krishan-singh , #1229
* [x] `org.carlspring.strongbox.services.MavenChecksumServiceTest`
* [x] `org.carlspring.strongbox.services.ArtifactMetadataServiceReleasesTest` @yugander-krishan-singh , #1229
* [x] `org.carlspring.strongbox.artifact.generation.MavenArtifactGeneratorTest`
* [x] `org.carlspring.strongbox.artifact.locator.ArtifactDirectoryLocatorTest`
* [x] `org.carlspring.strongbox.io.ArtifactOutputStreamTest`
* [x] `org.carlspring.strongbox.cron.jobs.RegenerateMavenChecksumCronJobTestIT` (@ptirador)
* [x] `org.carlspring.strongbox.cron.jobs.RebuildMavenMetadataCronJobTestIT` (@pyashwanthkumar)
* [x] `org.carlspring.strongbox.cron.jobs.ClearTrashCronJobFromMaven2RepositoryTestIT` (@ptirador)
* [x] `org.carlspring.strongbox.cron.jobs.DownloadRemoteMavenIndexCronJobTestIT` (@ptirador)
* [x] `org.carlspring.strongbox.cron.jobs.RemoveTimestampedMavenSnapshotCronJobTestIT` (@ptirador)
* [x] `org.carlspring.strongbox.cron.jobs.RebuildMavenIndexesCronJobTestIT` (@ptirador)
* [x] `org.carlspring.strongbox.cron.jobs.CleanupExpiredArtifactsFromProxyRepositoriesCronJobTestIT` (@ptirador)
* [x] `org.carlspring.strongbox.cron.jobs.BaseCronJobWithMavenIndexingTestCase` (@ptirador)
* [x] `org.carlspring.strongbox.repository.group.BaseMavenGroupRepositoryComponentTest`
* [x] `org.carlspring.strongbox.repository.group.index.BaseMavenIndexGroupRepositoryComponentTest`  (@ptirador)
* [x] `org.carlspring.strongbox.repository.group.index.MavenIndexGroupRepositoryComponentOnDeleteTest`
* [x] `org.carlspring.strongbox.repository.group.index.MavenIndexGroupRepositoryComponentOnCreateTest`
* [x] `org.carlspring.strongbox.repository.group.metadata.MavenMetadataGroupRepositoryComponentOnUploadTest`
* [x] `org.carlspring.strongbox.repository.group.metadata.MavenMetadataGroupRepositoryComponentOnAddTest`
* [x] `org.carlspring.strongbox.repository.group.metadata.MavenMetadataGroupRepositoryComponentOnDeleteTest`
* [x] `org.carlspring.strongbox.storage.validation.groupid.MavenGroupIdLowercaseValidatorTest` (@pyashwanthkumar)
* [x] `org.carlspring.strongbox.storage.validation.artifactid.MavenArtifactIdLowercaseValidatorTest` (@ptirador)
* [x] `org.carlspring.strongbox.storage.validation.DefaultMavenArtifactCoordinateValidatorsTest` (@ptirador)
* [x] `org.carlspring.strongbox.storage.validation.version.MavenReleaseVersionValidatorTest` (@ptirador)
* [x] `org.carlspring.strongbox.storage.validation.version.MavenSnapshotVersionValidatorTest` (@ptirador)
* [x] `org.carlspring.strongbox.storage.indexing.StrongboxIndexerTest` (@ptirador)
* [x] `org.carlspring.strongbox.storage.indexing.Maven2ProxyRepositoryTest`
* [x] `org.carlspring.strongbox.storage.indexing.RepositoryIndexerTest` (@ptirador)
* [x] `org.carlspring.strongbox.providers.io.BaseMavenMetadataExpirationTest` (@ptirador)
* [x] `org.carlspring.strongbox.providers.io.MavenMetadataExpirationProxyCaseTest` (@ptirador)
* [x] `org.carlspring.strongbox.providers.io.MavenMetadataExpirationSingleGroupCaseTest` (@ptirador)
* [x] `org.carlspring.strongbox.providers.io.MavenMetadataExpirationMultipleGroupCaseTest` (@ptirador)
* [x] `org.carlspring.strongbox.providers.repository.MavenGroupRepositoryProviderTest` (@ptirador)
* [x] `org.carlspring.strongbox.providers.repository.RetryDownloadArtifactWithUnsupportedRangeRequestTest`
* [x] `org.carlspring.strongbox.providers.repository.RetryDownloadArtifactTestBase`
* [x] `org.carlspring.strongbox.providers.repository.ParallelDownloadRemoteArtifactTest`
* [x] `org.carlspring.strongbox.providers.repository.RetryDownloadArtifactWithPermanentFailureStartingAtSomePointTest`
* [x] `org.carlspring.strongbox.providers.repository.WhenRepositoryIsNotAliveDontCleanExpiredArtifactsTestIT`
* [x] `org.carlspring.strongbox.providers.repository.BaseLocalStorageProxyRepositoryExpiredArtifactsCleanerTest`
* [x] `org.carlspring.strongbox.providers.repository.RetryDownloadArtifactWithSingleFailureAtSomePointTest`
* [x] `org.carlspring.strongbox.providers.repository.WhenRepositoryIsAliveCleanExpiredArtifactsTestIT`
* [x] `org.carlspring.strongbox.providers.repository.MavenProxyRepositoryProviderTestIT` (@pyashwanthkumar)
* [x] `org.carlspring.strongbox.providers.search.MavenIndexerSearchProviderTest` (@ptirador)
* [x] `org.carlspring.strongbox.providers.layout.Maven2LayoutProviderTest`
* [x] `org.carlspring.strongbox.testing.TestCaseWithMavenArtifactGenerationAndIndexing`(@ptirador)
* [x] `org.carlspring.strongbox.testing.MavenTestCaseWithArtifactGeneration`(@ptirador)
* [x] `org.carlspring.strongbox.resource.ResourcesBooterTest`
* [x] `org.carlspring.strongbox.dependency.snippet.SBTDependencyFormatterTest`
* [x] `org.carlspring.strongbox.dependency.snippet.IvyDependencyFormatterTest`
* [x] `org.carlspring.strongbox.dependency.snippet.MavenDependencyFormatterTest` (@ptirador)
* [x] `org.carlspring.strongbox.dependency.snippet.GradleDependencyFormatterTest`
* [x] `org.carlspring.strongbox.dependency.snippet.LeiningenDependencyFormatterTest`

# Task Relationships

This task:
* Is a sub-task of #839.
* Is a Follow-up of #1083 , #1084.
* Depends on #1191 , #1198.

# Useful Links

* [Writing Tests: Getting started](https://strongbox.github.io/developer-guide/writing-tests.html)
* [Writing Tests: JUnit5](https://strongbox.github.io/developer-guide/junit-user-guide.html)

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @sbespalov
  * @fuss86

**Comment by carlspring** [comment]
On hold. Looks like we'll need to polish up the annotations a bit first.

**Comment by carlspring** [comment]
Anyone wishing to help, could now feel free to join in. Please, have a look at the linked issues and pull requests to get a better idea of how to do things.

**Comment by carlspring** [comment]
@yugander-krishan-singh: Would you like to pick `MavenMetadataExpirationSingleGroupCaseTest` next?

**Comment by evrimnurcelik** [comment]
Can I work on RemoveTimestampedMavenSnapshotCronJobTestIT and RebuildMavenIndexesCronJobTestIT?

**Comment by carlspring** [comment]
@evrimnurcelik : Yes, please!

**Comment by carlspring** [comment]
@evrimnurcelik : 
Notes:

To enable the (deprecated) Maven Indexer, use `-Dmaven.indexer.enabled=true -Dspring.profiles.active=test`.

**Comment by yugander-krishan-singh** [comment]
@carlspring I am looking into `ArtifactManagementServiceImplTest` and after that I can pick up `MavenMetadataExpirationSingleGroupCaseTest`

**Comment by carlspring** [comment]
Okay, @yugander-krishan-singh. I've updated the list. Please, feel free to proceed!

**Comment by fuss86** [comment]
picking  `org.carlspring.strongbox.repository.group.BaseMavenGroupRepositoryComponentTest`

**Comment by pyashwanthkumar** [comment]
@carlspring  Can I pick ```org.carlspring.strongbox.cron.jobs.RebuildMavenMetadataCronJobTestIT```?

**Comment by carlspring** [comment]
@pyashwanthkumar ,

Yes, please, go ahead!

**Comment by Gcolon021** [comment]
@carlspring Can I pick `org.carlspring.strongbox.services.ConfigurationManagementServiceImplTest`?

**Comment by carlspring** [comment]
@Gcolon021 ,

Yes, of course! Have fun! :)

**Comment by ptirador** [comment]
Hi! I have one question:
In some tests, there's artifact creation with snapshot:
```

createTimestampedSnapshotArtifact(getRepositoryBasedir(STORAGE0, "rmmcjtit-snapshots").getAbsolutePath(),
                                  "org.carlspring.strongbox",
                                  "strongbox-metadata-one",
                                  "2.0",
                                  "jar",
                                  CLASSIFIERS,
                                  5);
```

This method is equivalent to this annotation?
```
@MavenTestArtifact(repositoryId = "rmmcjtit-snapshots",
                   id = "org.carlspring.strongbox:strongbox-metadata-one",
                   versions = { "2.0-20190701.202015-1",
                                "2.0-20190701.202101-2",
                                "2.0-20190701.202203-3",
                                "2.0-20190701.202311-4",
                                "2.0-20190701.202601-5" },
                   classifiers = { "javadoc",
                                   "sources",
                                   "source-release" })
Path artifact1
```

Or there's another simpler way to create an artifact with a specific number of builds?

**Comment by carlspring** [comment]
@ptirador : Yes, this is correct! We need to change this everywhere possible.

**Comment by pyashwanthkumar** [comment]
@carlspring  I am picking this MavenProxyRepositoryProviderTestIT.

**Comment by carlspring** [comment]
@pyashwanthkumar : Excellent! Thank you! :)

**Comment by ptirador** [comment]
@sbespalov / @carlspring, I have another question:
I'm trying to generate 2 artifacts in the same test method, both with the same version, but one has a classifier "sources" and the other not.
```
@MavenTestArtifact(repositoryId = "mdft-releases",
                   id = "org.carlspring.strongbox:maven-snippet",
                   versions = { "1.0" })
Path artifactPath,
@MavenTestArtifact(repositoryId = "mdft-releases",
                   id = org.carlspring.strongbox:maven-snippet",
                   versions = { "1.0" },
                   classifiers = { "sources" })
Path artifact2Path)
```

The result is that only the first artifact is generated, but not the second with classifier:
![](http://oi64.tinypic.com/15hd7a.jpg)

Any ideas?

**Comment by carlspring** [comment]
@ptirador ,

I think it's because you're using the same version. If you use the `@MavenTestArtifact` annotation and specify `classifiers = { "sources" }`, it should also create a `jar`, if I recall correctly.

**Comment by carlspring** [comment]
@sbespalov : Is my theory right?

**Comment by ptirador** [comment]
@carlspring : If I try what you said, just one annotation and specify one classifier, it's generating an artifact BUT as equal as my screenshot above: `maven-snippet-1.0.jar`
It's like is not taking into account the annotation attribute `classifiers`...

**Comment by sbespalov** [comment]
> @sbespalov : Is my theory right?

yes, it should like this, with one annotation

**Comment by sbespalov** [comment]
@ptirador, you can try to debug how the `MavenArtifactGeneratorStrategy` works in your case so that you can figure out why you have what it is :)

**Comment by ptirador** [comment]
Okay, here's what I saw:
In class `MavenArtifactGeneratorStrategy`, we have this:
```
Object classifiers = attributesMap.get("classifiers");
```

`attributesMap` is a `Map<String, Object>` with 8 attributes:

![](http://oi67.tinypic.com/dqfxae.jpg)

Here we can seee that there's no `classifiers` key. If we trace back to see how `attributesMap` is filled, we arrive to `ArtifactManagementTestExecutionListener`class, where we can see this code:
```
TestArtifact testArtifact = AnnotatedElementUtils.findMergedAnnotation(parameterContext.getParameter(), TestArtifact.class);
        
Map<String, Object> attributesMap = Arrays.stream(parameterContext.getParameter().getAnnotations())
                                          .flatMap(a -> AnnotatedElementUtils.getMetaAnnotationTypes(parameterContext.getParameter(),
                                                                                                     a.annotationType()
                                                                                                      .getName())
                                                                             .stream())
                                          .flatMap(a -> AnnotatedElementUtils.getMergedAnnotationAttributes(parameterContext.getParameter(),
                                                                                                            a)
                                                                             .entrySet()
                                                                             .stream())
                                          .collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue(), (k1, k2) -> k1));
```

I don't really understand what it does here, but it's like is taking `TestArtifact` attributes (it has 8, as same as the screenshot above) instead of `MavenTestArtifact` ones?

**Comment by sbespalov** [comment]
yeah, this seems like a bug
I will have a look tomorrow

**Comment by ptirador** [comment]
Okay, thanks @sbespalov :)

**Comment by sbespalov** [comment]
@ptirador as a workaround you can try use it with shortcut annotations. Please have a look at `ArtifactMetadataServiceSnapshotsTest.MavenArtifactWithClassifiers`

**Comment by sbespalov** [comment]
@ptirador @carlspring I've fixed the `classifiers` attribute:
https://github.com/strongbox/strongbox/pull/1372

**Comment by carlspring** [comment]
Thanks! I've merged it into the `master`.

**Comment by pyashwanthkumar** [comment]
@carlspring / @sbespalov  

I would also like to pick up other test cases while I am working on my previous ones. The one that I am looking at right now is this test case "MavenGroupIdLowercaseValidatorTest". In this test, do we want to replace instantiation of MavenArtifactCoordinates and place this inside the test case(after inserting test annotations) or am I looking at it wrong?

**Comment by carlspring** [comment]
@pyashwanthkumar : Have a look at some of the associated pull requests for examples of what needs to be done.

**Comment by ptirador** [comment]
> @carlspring / @sbespalov
> 
> I would also like to pick up other test cases while I am working on my previous ones. The one that I am looking at right now is this test case "MavenGroupIdLowercaseValidatorTest". In this test, do we want to replace instantiation of MavenArtifactCoordinates and place this inside the test case(after inserting test annotations) or am I looking at it wrong?

@pyashwanthkumar: Take a look at this [commit](https://github.com/strongbox/strongbox/pull/1370/commits/5de2a325078c599aa16534e0ebe6de33c91270c9) I just did.

**Comment by ptirador** [comment]
@Gcolon021 @yugander-krishan-singh @pyashwanthkumar @SergeyHydra
Are any of hoy still working on this?

**Comment by pyashwanthkumar** [comment]
@ptirador yes, I am still working on MavenGroupIdLowercaseValidatorTest and MavenProxyRepositoryProviderTestIT :)

**Comment by carlspring** [comment]
@ptirador : There's a PR for one of the tests in #1360.

I've removed @Gcolon021 and @SergeyHydra from the list of tests, as they're not working on them.

**Comment by SergeyHydra** [comment]
Hm, hi! I'm still working on test.

**Comment by carlspring** [comment]
Ah, okay, sorry, @SergeyHydra ! Which one was it?

**Comment by ptirador** [comment]
I've already added it @carlspring  ;)
It was `BaseMavenIndexGroupRepositoryComponentTest`.

**Comment by SergeyHydra** [comment]
Wow! So what are left undone?

**Comment by carlspring** [comment]
Well, the ones that have @ptirador 's name against them, but are not ticked, are already things he's either done in his latest pull request, or will include in it.

Have a look for any remaining ones, or ping us on the channel for new tasks. There's plenty more tasks to be done! :)

**Comment by carlspring** [comment]
@pyashwanthkumar,

Are you currently still working on:
* `org.carlspring.strongbox.storage.validation.groupid.MavenGroupIdLowercaseValidatorTest `
* `org.carlspring.strongbox.providers.repository.MavenProxyRepositoryProviderTestIT`

If you're not, then perhaps @ptirador could have a look at them as well and we can wrap this up?

**Comment by pyashwanthkumar** [comment]
> @pyashwanthkumar,
> 
> Are you currently still working on:
> 
> * `org.carlspring.strongbox.storage.validation.groupid.MavenGroupIdLowercaseValidatorTest `
> * `org.carlspring.strongbox.providers.repository.MavenProxyRepositoryProviderTestIT`
> 
> If you're not, then perhaps @ptirador could have a look at them as well and we can wrap this up?

@carlspring I finished working on org.carlspring.strongbox.storage.validation.groupid.MavenGroupIdLowercaseValidatorTest.

I am almost done with org.carlspring.strongbox.providers.repository.MavenProxyRepositoryProviderTestIT.

If I am unable to finish it by the end of day Saturday, I will pass this on to @ptirador.

**Comment by carlspring** [comment]
Thanks! Take your time! :)

Just checking...

**Comment by carlspring** [comment]
A massive thank you to everybody who helped with this extremely large refactoring!

**Comment by steve-todorov** [comment]
Thanks guys! This was a huge task indeed and we deeply appreciate all of your efforts and contributions! :)
